File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
    let maxf = 0;
    let map = {};
    let left = 0;
    let longest = 0;
    
    for (let right = 0; right < s.length; right++) {
        map[s[right]] = s[right] in map ? map[s[right]] + 1 : 1;
        maxf = Math.max(maxf, map[s[right]]);
        
        while (right - left + 1 - maxf > k) {
            map[s[left]] -= 1
            left += 1;
        }
        
        longest = Math.max(longest, right - left + 1);
        
    }
    
    return longest;
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        Map<Character, Integer> count = new HashMap<>();
        int l = 0;
        int maxFreq = 0;
        int longest = 0;

        for (int r = 0; r < s.length(); r++) {
            count.put(s.charAt(r), count.getOrDefault(s.charAt(r), 0) + 1);
            maxFreq = Math.max(maxFreq, count.get(s.charAt(r)));

            while (r - l + 1 - maxFreq > k) {
                count.put(s.charAt(l), count.get(s.charAt(l))-1);
                l++;
            }

            longest = Math.max(longest, r - l + 1);
        }

        return longest;
    }
}